### ISSUE TITLE
Steps take extremely long with great processor and low buffer-size

### ISSUE THREAD

**Issue by tillmusshoff** [Unity-Technologies/ml-agents#2323]
Hi! As the title says the steps take very long in the training process. 1000 steps are between 2-3 minutes.  

I am using Anaconda and run a 12-core i9-7920X 2,90GHz CPU. Reducing the buffer-size from 80000 to 8000 only made very small improvements. The CPU is only used up to 20% during training while I still use the web. I run 16 environments at once with --num-envs=16 and --no-graphics.

You can see my Hyperparameters and Observation- as well as Action-space in the screenshots provided.

The training-progress is great after 500k steps, however that takes very long and should go fast I believe. I run a more simple version on my old MacBook which wasn't that much slower.

Thanks for your help!

![screen2](https://user-images.githubusercontent.com/53266983/61802447-bc295780-ae30-11e9-82b3-ffacbd541376.png)
![screen1](https://user-images.githubusercontent.com/53266983/61802449-bc295780-ae30-11e9-90f5-881115ee8558.png)

**Comment by harperj** [comment]
Hi @tillmusshoff -- it's a bit hard to know why your environment steps are taking so long without knowing the specifics of how your environment is set up.  You might consider the time scale settings you've used within the Academy config.  One thing that can currently slow down training with parallel environments is also when some steps take longer than others (since all of the environments step in parallel but wait for the others).  This type of slowdown should be improved in the v0.9 release which is coming soon.

**Comment by tillmusshoff** [comment]
Thanks for you reply. I attached the Academy config as a screenshot. I also tested the training again with only one enivornment. The time for 1000 steps got reduced from 120 seconds to about 30 seconds. However only 10% of the processing power is being used. Is there a way to increase this drastically? I run a 1000$ processor thats not being used to it's full potential. 

Also would you recommend training with multiple environments and longer timesteps or one environment with shorter timesteps? 

Thanks a lot!

![c4cce7a5602eed2daec256a0780093e2](https://user-images.githubusercontent.com/53266983/61934046-af6b4780-af87-11e9-9011-214e663f4dd6.png)

**Comment by AcelisWeaven** [comment]
Hi @tillmusshoff ,

I saw your post and ran some tests to see what would happen, and indeed it seemed faster.
But I don't think it really is, the reason for this is that 1 step = 1 tick in Unity. So if you have one agent, in one step your agent will train on one observation, whereas with ten agents it'll train on ten observations (thus, it should learn faster)

Anyway I wanted to benchmark it, so here's a graph from a training session on a virtual server with 1 vCPU, where I changed my number of agents. I'm using PR #1975, where an agent learn to play against himself. ELO rating is a good metric in my environment, but I don't know what metric you could use to measure the "learning speed".

![image](https://user-images.githubusercontent.com/6735195/61949367-f6b6ff80-afaa-11e9-8dc9-20a03366f6b6.png)

From this graph, my ELO increased like this:

| Session | From | To | Time | ELO increase | ELO points/minute |
| :--- | :---: | :---: | :---: | :---: | ---: |
| 20 agents (first run) | 1529 | 1599 | 95 min. | +70 | **+0,73** |
| 2 agents (second run)  | 1599 | 1604 | 15 min. | +5 | **+0.33** |
| 20 agents (third run)  | 1604 | 1666 | 90 min. | +62 | **+0.69** |

In my case, running 10x more agents make the training twice faster. I'll need more tests to see if I can find a good number of agents to maximise this ratio.

Also (I don't have metrics for this, sorry!), switching my virtual server to a 8 vCPU instance actually reduced the training speed a lot! And I don't know why.

Unity team, do you have some insights?

**Comment by tillmusshoff** [comment]
Thanks for your research @AcelisWeaven. I just found out, that out of my 12-cores (24-logic-cores) only two are used. That seems odd, doesn't it?

![96a126620a3b35ee1a12f727d7282ecd](https://user-images.githubusercontent.com/53266983/61952813-28cd5f00-afb5-11e9-8189-b62d0ec7e4a1.png)

**Comment by roboserg** [comment]
I also noticed that adding one agent in an inference mode (while training the other) adds around 10 seconds to the 1000 steps of training. In other words it takes 10 seconds for only one single agent to train. If you add a second agent that is NOT training (only uses inference) the time jumps to 20s per 1000 training steps. I find it really weird, as the impact of the inference should theoretically not be so huge.
